core/builder.go

```diff
@@ -62,53 +86,71 @@ func defaultRepo(dstore ds.ThreadSafeDatastore) (repo.Repo, error) {
 	}, nil
 }
 
-func (nb *NodeBuilder) Online() *NodeBuilder {
-	nb.online = true
-	return nb
-}
+func NewNode(ctx context.Context, cfg *BuildCfg) (*IpfsNode, error) {
+	if cfg == nil {
+		cfg = new(BuildCfg)
+	}
 
-func (nb *NodeBuilder) Offline() *NodeBuilder {
-	nb.online = false
-	return nb
-}
+	err := cfg.fillDefaults()
+	if err != nil {
+		return nil, err
+	}
 
-func (nb *NodeBuilder) SetRouting(ro RoutingOption) *NodeBuilder {
-	nb.routing = ro
-	return nb
-}
+	n := &IpfsNode{
+		mode:      offlineMode,
+		Repo:      cfg.Repo,
+		ctx:       ctx,
+		Peerstore: peer.NewPeerstore(),
+	}
+	if cfg.Online {
+		n.mode = onlineMode
+	}
 
-func (nb *NodeBuilder) SetHost(ho HostOption) *NodeBuilder {
-	nb.peerhost = ho
-	return nb
-}
+	// TODO: this is a weird circular-ish dependency, rework it
+	n.proc = goprocessctx.WithContextAndTeardown(ctx, n.teardown)
 
-func (nb *NodeBuilder) SetRepo(r repo.Repo) *NodeBuilder {
-	nb.repo = r
-	return nb
-}
+	success := false
+	defer func() {
+		if !success {
+			n.teardown()
+		}
+	}()
 
-func (nb *NodeBuilder) NilRepo() *NodeBuilder {
-	nb.nilrepo = true
-	return nb
-}
+	// setup local peer ID (private key is loaded in online setup)
+	if err := n.loadID(); err != nil {
+		return nil, err
+	}
 
-func (nb *NodeBuilder) Build(ctx context.Context) (*IpfsNode, error) {
-	if nb.built {
-		return nil, ErrAlreadyBuilt
+	n.Blockstore, err = bstore.WriteCached(bstore.NewBlockstore(n.Repo.Datastore()), kSizeBlockstoreWriteCache)
+	if err != nil {
+		return nil, err
 	}
-	nb.built = true
-	if nb.repo == nil {
-		var d ds.Datastore
-		d = ds.NewMapDatastore()
-		if nb.nilrepo {
-			d = ds.NewNullDatastore()
-		}
-		r, err := defaultRepo(dsync.MutexWrap(d))
-		if err != nil {
+
+	if cfg.Online {
+		do := setupDiscoveryOption(n.Repo.Config().Discovery)
+		if err := n.startOnlineServices(ctx, cfg.Routing, cfg.Host, do); err != nil {
 			return nil, err
 		}
-		nb.repo = r
+	} else {
+		n.Exchange = offline.Exchange(n.Blockstore)
 	}
-	conf := standardWithRouting(nb.repo, nb.online, nb.routing, nb.peerhost)
-	return NewIPFSNode(ctx, conf)
+
+	n.Blocks = bserv.New(n.Blockstore, n.Exchange)
+	if err != nil {
+		return nil, err
+	}
+
+	n.DAG = dag.NewDAGService(n.Blocks)
+	n.Pinning, err = pin.LoadPinner(n.Repo.Datastore(), n.DAG)
+	if err != nil {
+		// TODO: we should move towards only running 'NewPinner' explicity on
+		// node init instead of implicitly here as a result of the pinner keys
+		// not being found in the datastore.
+		// this is kinda sketchy and could cause data loss
+		n.Pinning = pin.NewPinner(n.Repo.Datastore(), n.DAG)
+	}
+	n.Resolver = &path.Resolver{DAG: n.DAG}
+
+	success = true
```

Comment:
this function is a scary still, and the `success/defer` stuff is not easy to modify safely (error prone).

how about this:

```go
func cfgDefaults(cfg *BuildCfg) (*BuildCfg, error) {
  if cfg == nil {
    cfg = new(BuildCfg)
  }

  err := cfg.fillDefaults()
  if err != nil {
    return nil, err
  }    
}

func NewNode(ctx context.Context, cfg *BuildCfg) (*IpfsNode, error) {
  cfg, err := cfgDefaults(cfg)
  if err != nil {
    return nil, err
  }

  n := &IpfsNode{
    mode:      offlineMode,
    Repo:      cfg.Repo,
    ctx:       ctx,
    Peerstore: peer.NewPeerstore(),
  }
  if cfg.Online {
    n.mode = onlineMode
  }

  // TODO: this is a weird circular-ish dependency, rework it
  n.proc = goprocessctx.WithContextAndTeardown(ctx, n.teardown)
  if err := tryNodeSetup(ctx, cfg, n); err != nil {
    n.Close() 
    return nil, err
  }

  return n, nil
}

func tryNodeSetup(ctx context.Context, cfg *BuildCfg, n *IpfsNode) error {
  // setup local peer ID (private key is loaded in online setup)
  if err := n.loadID(); err != nil {
    return nil, err
  }

  bs := bstore.NewBlockstore(n.Repo.Datastore())
  n.Blockstore, err = bstore.WriteCached(bs, kSizeBlockstoreWriteCache)
  if err != nil {
    return nil, err
  }

  if cfg.Online {
    do := setupDiscoveryOption(n.Repo.Config().Discovery)
    if err := n.startOnlineServices(ctx, cfg.Routing, cfg.Host, do); err != nil {
      return nil, err
    }
  } else {
    n.Exchange = offline.Exchange(n.Blockstore)
  }

  n.Blocks = bserv.New(n.Blockstore, n.Exchange)
  n.DAG = dag.NewDAGService(n.Blocks)
  n.Pinning, err = pin.LoadPinner(n.Repo.Datastore(), n.DAG)
  if err != nil {
    // TODO: we should move towards only running 'NewPinner' explicity on
    // node init instead of implicitly here as a result of the pinner keys
    // not being found in the datastore.
    // this is kinda sketchy and could cause data loss
    n.Pinning = pin.NewPinner(n.Repo.Datastore(), n.DAG)
  }
  n.Resolver = &path.Resolver{DAG: n.DAG}
  return nil
}
```

Note:  call `n.Close()`, or even `n.proc.Close()`. do not call `n.teardown()` directly!! or it may be called twice and break semantics (maybe panic)